python/vast/vast/vast.py

```diff
@@ -37,78 +102,45 @@ async def start():
-                        # hashable. The string representation is the next best
-                        # thing to determine uniqueness.
-                        batches[batch.schema.to_string()].append(batch)
-                except StopIteration:
-                    logger.debug(f"read {num_batches}/{num_rows} batches/rows")
-                    num_batches = 0
-                    num_rows = 0
-        except pa.ArrowInvalid:
-            logger.debug("completed processing stream of record batches")
-        result = defaultdict(list)
-        for (_, batches) in batches.items():
-            table = pa.Table.from_batches(batches)
-            name = vast.utils.arrow.name(table.schema)
-            result[name].append(table)
-        return result
-
     @staticmethod
-    async def export_rows(
-        expression: str, mode: ExportMode, limit: int = -1
-    ) -> AsyncIterable[Dict[str, Any]]:
-        """Get a row wise view of the data
-
-        This method does not provide type information because it is using json
-        as export format. It is a temporary workaround for asynchronicity issues
-        with PyArrow, but we plan to have all VAST Python exports properly typed
-        and backed by Arrow."""
+    async def export(
+        expression: str, mode: ExportMode = ExportMode.HISTORICAL, limit: int = 100
+    ) -> AsyncIterable[TableSlice]:
+        """Executes a VAST and receives results as Arrow Tables."""
         if limit == 0:
             return
-        args = {}
-        match mode:
-            case ExportMode.CONTINUOUS:
-                args["continuous"] = True
-            case ExportMode.UNIFIED:
-                args["unified"] = True
-            case ExportMode.HISTORICAL:
-                pass
-        if limit > 0:
-            args["max_events"] = limit
-        proc = await CLI().export(**args).json(expression).exec()
-        while True:
-            if proc.stdout.at_eof():
-                break
-            line = await proc.stdout.readline()
-            if line.strip() == b"":
-                continue
-            line_dict = json.loads(line)
-            assert isinstance(line_dict, Dict)
-            yield line_dict
-
-        _, stderr = await proc.communicate()
-        if proc.returncode != 0:
-            logger.error(stderr.decode())
+        cmd = CLI().export(**VAST._export_args(mode, limit))
+        if expression == "":
+            cmd = cmd.arrow()
+        else:
+            cmd = cmd.arrow(expression)
+        proc = cmd.sync_exec()
+
+        def log():
+            id = secrets.token_hex(6)
+            export_str = f"export({expression}, {mode}, {limit})"
+            logger.debug(f"[export-{id}] Start logging for {export_str}")
+            for line in iter(proc.stderr.readline, b""):
+                logger.debug(f"[export-{id}] {line.decode().strip()}")
+            logger.debug(f"[export-{id}] Stop logging")
+
+        # TODO: don't use default thread pool here as it won't scale
```

Comment:
Good question! the default executor is [`concurrent.futures.ThreadPoolExecutor`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor). This has a default max thread (proc on the machine x5). It will not grow indefinitely and we might quickly be limited if we run multiple exports in parallel. Using a thread here really defeats the purpose of asyncio 😢. There is a [workaround](https://github.com/cloudfuse-io/lambdatization/blob/eb29b12ce0363415bfafca2d09ec605d705dd814/docker/clickhouse/lambda-handler.py#L18-L63) that uses selectors and would make it possible to do all the logging on a single thread, but I really don't want to over-engineer right away + make this PR to big

Reply:
TIL. Can you change the comment to something like
```suggestion
        # TODO: don't use default thread pool here as this method 
        # will block one thread per running export, risking exhaustion of 
        # the pool if we run many in parallel.
```